File: dotnet/src/Tyouqu.Database.Abstractions/SqlSafetyAnalyzer.cs
using System.Text;

namespace Tyouqu.Database.Abstractions;

public static class SqlSafetyAnalyzer
{
    public static string Normalize(string sql)
    {
        var builder = new StringBuilder(sql.Length);
        var inSingleQuote = false;
        var inDoubleQuote = false;
        var inLineComment = false;
        var inBlockComment = false;

        for (var i = 0; i < sql.Length; i++)
        {
            var current = sql[i];
            var next = i + 1 < sql.Length ? sql[i + 1] : '\0';

            if (inLineComment)
            {
                if (current is '\r' or '\n')
                {
                    inLineComment = false;
                    builder.Append(' ');
                }
                continue;
            }

            if (inBlockComment)
            {
                if (current == '*' && next == '/')
                {
                    inBlockComment = false;
                    i++;
                    builder.Append(' ');
                }
                continue;
            }

            if (!inSingleQuote && !inDoubleQuote && current == '-' && next == '-')
            {
                inLineComment = true;
                i++;
                continue;
            }

            if (!inSingleQuote && !inDoubleQuote && current == '/' && next == '*')
            {
                inBlockComment = true;
                i++;
                continue;
            }

            if (!inDoubleQuote && current == '\'')
            {
                inSingleQuote = !inSingleQuote;
                builder.Append(' ');
                continue;
            }

            if (!inSingleQuote && current == '"')
            {
                inDoubleQuote = !inDoubleQuote;
                builder.Append(' ');
                continue;
            }

            builder.Append(inSingleQuote || inDoubleQuote ? ' ' : char.ToLowerInvariant(current));
        }

        return string.Join(' ', builder.ToString().Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries));
    }

    public static bool IsFullTableUpdate(string normalizedSql)
    {
        return normalizedSql.StartsWith("update ", StringComparison.Ordinal) &&
               !ContainsWhereClause(normalizedSql);
    }

    public static bool IsFullTableDelete(string normalizedSql)
    {
        return (normalizedSql.StartsWith("delete from ", StringComparison.Ordinal) ||
                normalizedSql == "delete" ||
                normalizedSql.StartsWith("delete ", StringComparison.Ordinal)) &&
               !ContainsWhereClause(normalizedSql);
    }

    private static bool ContainsWhereClause(string normalizedSql)
    {
        return normalizedSql.Split(' ', StringSplitOptions.RemoveEmptyEntries).Contains("where");
    }
}

File: java/src/main/java/com/tyouqu/database/SqlSafetyAnalyzer.java
package com.tyouqu.database;

import java.util.Arrays;
import java.util.Locale;

public final class SqlSafetyAnalyzer {
    private SqlSafetyAnalyzer() {
    }

    public static String normalize(String sql) {
        StringBuilder builder = new StringBuilder(sql.length());
        boolean inSingleQuote = false;
        boolean inDoubleQuote = false;
        boolean inLineComment = false;
        boolean inBlockComment = false;

        for (int i = 0; i < sql.length(); i++) {
            char current = sql.charAt(i);
            char next = i + 1 < sql.length() ? sql.charAt(i + 1) : '\0';

            if (inLineComment) {
                if (current == '\r' || current == '\n') {
                    inLineComment = false;
                    builder.append(' ');
                }
                continue;
            }
            if (inBlockComment) {
                if (current == '*' && next == '/') {
                    inBlockComment = false;
                    i++;
                    builder.append(' ');
                }
                continue;
            }
            if (!inSingleQuote && !inDoubleQuote && current == '-' && next == '-') {
                inLineComment = true;
                i++;
                continue;
            }
            if (!inSingleQuote && !inDoubleQuote && current == '/' && next == '*') {
                inBlockComment = true;
                i++;
                continue;
            }
            if (!inDoubleQuote && current == '\'') {
                inSingleQuote = !inSingleQuote;
                builder.append(' ');
                continue;
            }
            if (!inSingleQuote && current == '"') {
                inDoubleQuote = !inDoubleQuote;
                builder.append(' ');
                continue;
            }

            builder.append(inSingleQuote || inDoubleQuote ? ' ' : Character.toLowerCase(current));
        }

        return String.join(" ", builder.toString().trim().split("\\s+"));
    }

    public static boolean isFullTableUpdate(String normalizedSql) {
        return normalizedSql.startsWith("update ") && !containsWhereClause(normalizedSql);
    }

    public static boolean isFullTableDelete(String normalizedSql) {
        return (normalizedSql.startsWith("delete from ") || normalizedSql.equals("delete") || normalizedSql.startsWith("delete "))
            && !containsWhereClause(normalizedSql);
    }

    private static boolean containsWhereClause(String normalizedSql) {
        return Arrays.asList(normalizedSql.toLowerCase(Locale.ROOT).split("\\s+")).contains("where");
    }
}
